nautobot/extras/health_checks.py

```diff
@@ -65,10 +66,32 @@ def check_status(self):
                 obj.title = "newtest"
                 obj.save()
                 obj.delete()
-        except IntegrityError:
-            raise ServiceReturnedUnexpectedResult("Integrity Error")
-        except DatabaseError:
-            raise ServiceUnavailable("Database error")
+        except IntegrityError as e:
+            self.add_error(ServiceReturnedUnexpectedResult("Integrity Error"), e)
+        except DatabaseError as e:
+            self.add_error(ServiceUnavailable("Database error"), e)
+
+
+class MigrationsBackend(NautobotHealthCheckBackend):
+    """Check whether all migrations have been applied."""
+
+    metric = Gauge(
+        "health_check_migrations_info",
+        "State of migrations backend",
+        multiprocess_mode=NautobotHealthCheckBackend.MULTIPROCESS_MODE,
+    )
+
+    def check_status(self):
+        db_alias = getattr(settings, "HEALTHCHECK_MIGRATIONS_DB", DEFAULT_DB_ALIAS)
+        try:
+            executor = MigrationExecutor(connections[db_alias])
+            plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
```

Comment:
Building the migration plan can be CPU intensive. Is there a way we can query the `django_migrations` table and find out if any of the migrations aren't applied?

Reply:
The `django_migrations` table only records migrations that *have been* applied, so we have to rely on Django machinery to tell us which migrations *should be* applied.

I tried the following alternate implementation:

```diff
@@ -84,10 +85,11 @@ class MigrationsBackend(NautobotHealthCheckBackend):
     def check_status(self):
         db_alias = getattr(settings, "HEALTHCHECK_MIGRATIONS_DB", DEFAULT_DB_ALIAS)
         try:
-            executor = MigrationExecutor(connections[db_alias])
-            plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
-            if plan:
-                self.add_error(ServiceUnavailable("There are migrations not yet applied"))
+            loader = MigrationLoader(connections[db_alias], ignore_no_migrations=True)
+            for app_name, final_migration in loader.graph.leaf_nodes():
+                if (app_name, final_migration) not in loader.applied_migrations:
+                    self.add_error(ServiceUnavailable("There are migrations not yet applied"))
+                    break
         except DatabaseError as e:
             self.add_error(ServiceUnavailable("Database is not ready"), e)
         except Exception as e:
```

but `cProfile` shows little performance difference between the two:

### Original implementation

```
         74090 function calls (74065 primitive calls) in 0.115 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     19/1    0.000    0.000    0.115    0.115 {built-in method builtins.exec}
        1    0.000    0.000    0.115    0.115 <string>:1(<module>)
        1    0.000    0.000    0.114    0.114 health_checks.py:84(check_status)
        1    0.000    0.000    0.113    0.113 executor.py:16(__init__)
        1    0.000    0.000    0.113    0.113 loader.py:43(__init__)
        1    0.000    0.000    0.113    0.113 loader.py:207(build_graph)
        1    0.001    0.001    0.095    0.095 loader.py:68(load_disk)
       18    0.000    0.000    0.045    0.002 loader.py:108(<setcomp>)
      368    0.001    0.000    0.044    0.000 pkgutil.py:110(iter_modules)
      368    0.002    0.000    0.043    0.000 pkgutil.py:144(_iter_file_finder_modules)
       18    0.000    0.000    0.034    0.002 __init__.py:133(reload)
       36    0.026    0.001    0.026    0.001 {built-in method posix.listdir}
      164    0.021    0.000    0.021    0.000 {built-in method posix.stat}
       18    0.000    0.000    0.019    0.001 <frozen importlib._bootstrap>:583(_exec)
       18    0.000    0.000    0.018    0.001 <frozen importlib._bootstrap_external>:837(exec_module)
       18    0.000    0.000    0.018    0.001 <frozen importlib._bootstrap_external>:909(get_code)
       38    0.001    0.000    0.016    0.000 <frozen importlib._bootstrap>:890(_find_spec)
      128    0.000    0.000    0.016    0.000 <frozen importlib._bootstrap_external>:135(_path_stat)
       38    0.000    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1399(find_spec)
       38    0.000    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1367(_get_spec)
       38    0.001    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1498(find_spec)
       18    0.001    0.000    0.015    0.001 <frozen importlib._bootstrap_external>:1029(get_data)
      390    0.001    0.000    0.013    0.000 __init__.py:109(import_module)
      390    0.000    0.000    0.013    0.000 <frozen importlib._bootstrap>:1002(_gcd_import)
      390    0.002    0.000    0.012    0.000 <frozen importlib._bootstrap>:986(_find_and_load)
```


### Alternative implementation

```
         73804 function calls (73779 primitive calls) in 0.107 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     19/1    0.000    0.000    0.107    0.107 {built-in method builtins.exec}
        1    0.000    0.000    0.107    0.107 <string>:1(<module>)
        1    0.000    0.000    0.106    0.106 health_checks.py:85(check_status)
        1    0.000    0.000    0.105    0.105 loader.py:43(__init__)
        1    0.001    0.001    0.105    0.105 loader.py:207(build_graph)
        1    0.001    0.001    0.084    0.084 loader.py:68(load_disk)
       18    0.000    0.000    0.039    0.002 loader.py:108(<setcomp>)
      368    0.000    0.000    0.038    0.000 pkgutil.py:110(iter_modules)
      368    0.001    0.000    0.037    0.000 pkgutil.py:144(_iter_file_finder_modules)
       18    0.000    0.000    0.031    0.002 __init__.py:133(reload)
       37    0.024    0.001    0.024    0.001 {built-in method posix.listdir}
      164    0.018    0.000    0.018    0.000 {built-in method posix.stat}
       18    0.000    0.000    0.017    0.001 <frozen importlib._bootstrap>:583(_exec)
       38    0.001    0.000    0.016    0.000 <frozen importlib._bootstrap>:890(_find_spec)
       18    0.000    0.000    0.016    0.001 <frozen importlib._bootstrap_external>:837(exec_module)
       18    0.000    0.000    0.016    0.001 <frozen importlib._bootstrap_external>:909(get_code)
       38    0.000    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1399(find_spec)
       38    0.000    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1367(_get_spec)
       38    0.000    0.000    0.015    0.000 <frozen importlib._bootstrap_external>:1498(find_spec)
      128    0.000    0.000    0.014    0.000 <frozen importlib._bootstrap_external>:135(_path_stat)
       18    0.001    0.000    0.013    0.001 <frozen importlib._bootstrap_external>:1029(get_data)
      390    0.000    0.000    0.012    0.000 __init__.py:109(import_module)
      390    0.000    0.000    0.012    0.000 <frozen importlib._bootstrap>:1002(_gcd_import)
      390    0.002    0.000    0.011    0.000 <frozen importlib._bootstrap>:986(_find_and_load)
```

Reply:
It's the graph build that's CPU intensive. The majority of the time is spent in `loader.py:207(build_graph)` but a big percentage of that is loading the migrations from disk. Probably not worth worrying about until we get a few thousand more migration files 😄 but thank you for humoring me!

Reply:
Sure thing. It's been helpful in getting my hands dirty with `cProfile` a bit more. Relatedly, it *looks* like the majority of the `nautobot-server` startup time in general is spent purely in discovering and importing modules rather than doing any particularly expensive function calls. So that'll be interesting to try and optimize.